File: iOS/KinKitchen/KinKitchen/Features/Profile/EditDietaryProfileView.swift
//
//  EditDietaryProfileView.swift
//  KinKitchen
//
//  Created by Greg Hudler on 8/30/26.
//

import SwiftUI

struct EditDietaryProfileView: View {

    @Environment(\.dismiss) private var dismiss

    @State private var allergens: [Allergen] = []
    @State private var restrictions: [DietaryRestriction] = []
    @State private var preferences: [DietaryPreference] = []

    @State private var selectedAllergenIds: Set<UUID> = []
    @State private var selectedRestrictionIds: Set<UUID> = []
    @State private var selectedPreferenceIds: Set<UUID> = []

    @State private var originalAllergenIds: Set<UUID> = []
    @State private var originalRestrictionIds: Set<UUID> = []
    @State private var originalPreferenceIds: Set<UUID> = []

    @State private var isLoading = true
    @State private var isSaving = false
    @State private var errorMessage: String?

    var body: some View {
            ScrollView {
                VStack(alignment: .leading, spacing: KinSpacing.xLarge) {

                    if isLoading {

                        KinLoadingView()

                    } else {

                        dietarySection(
                            title: "Allergens",
                            items: allergens,
                            selectedIds: $selectedAllergenIds
                        )

                        dietarySection(
                            title: "Dietary Restrictions",
                            items: restrictions,
                            selectedIds: $selectedRestrictionIds
                        )

                        dietarySection(
                            title: "Dietary Preferences",
                            items: preferences,
                            selectedIds: $selectedPreferenceIds
                        )

                        if let errorMessage {
                            Text(errorMessage)
                                .font(KinTypography.body)
                                .foregroundStyle(KinColors.error)
                        }

                        KinPrimaryButton(
                            title: isSaving ? "Saving..." : "Save Changes",
                            color: KinColors.success
                        ) {
                            Task {
                                await saveChanges()
                            }
                        }
                        .disabled(isSaving)
                    }
                }
                .padding(KinSpacing.xLarge)
            }
            .background(KinColors.background)
            .navigationTitle("Edit Dietary Profile")
            .navigationBarTitleDisplayMode(.inline)
            .task {
                await loadDietaryInformation()
            }
    }

    private func dietarySection<Item: Identifiable>(
        title: String,
        items: [Item],
        selectedIds: Binding<Set<UUID>>
    ) -> some View where Item.ID == UUID {

        VStack(alignment: .leading, spacing: KinSpacing.medium) {

            Text(title)
                .font(KinTypography.title)
                .foregroundStyle(KinColors.primaryText)

            KinFlowLayout(spacing: KinSpacing.small) {
                ForEach(items) { item in
                    Button {
                        toggle(
                            id: item.id,
                            selectedIds: selectedIds
                        )
                    } label: {
                        KinChip(
                            title: itemName(item),
                            isSelected: selectedIds.wrappedValue.contains(item.id)
                        )
                    }
                    .buttonStyle(.plain)
                }
            }
        }
    }

    private func itemName<Item>(_ item: Item) -> String {
        if let allergen = item as? Allergen {
            return allergen.name
        }

        if let restriction = item as? DietaryRestriction {
            return restriction.name
        }

        if let preference = item as? DietaryPreference {
            return preference.name
        }

        return ""
    }

    private func toggle(
        id: UUID,
        selectedIds: Binding<Set<UUID>>
    ) {
        if selectedIds.wrappedValue.contains(id) {
            selectedIds.wrappedValue.remove(id)
        } else {
            selectedIds.wrappedValue.insert(id)
        }
    }

    @MainActor
    private func loadDietaryInformation() async {
        isLoading = true
        errorMessage = nil

        defer {
            isLoading = false
        }

        do {
            async let loadedAllergens = DietaryService.fetchAllergens()
            async let loadedRestrictions = DietaryService.fetchDietaryRestrictions()
            async let loadedPreferences = DietaryService.fetchDietaryPreferences()

            async let savedAllergens = DietaryService.fetchSelectedAllergens()
            async let savedRestrictions = DietaryService.fetchSelectedDietaryRestrictions()
            async let savedPreferences = DietaryService.fetchSelectedDietaryPreferences()

            allergens = try await loadedAllergens
            restrictions = try await loadedRestrictions
            preferences = try await loadedPreferences

            selectedAllergenIds = Set(try await savedAllergens)
            selectedRestrictionIds = Set(try await savedRestrictions)
            selectedPreferenceIds = Set(try await savedPreferences)

            originalAllergenIds = selectedAllergenIds
            originalRestrictionIds = selectedRestrictionIds
            originalPreferenceIds = selectedPreferenceIds

        } catch {
            errorMessage = error.localizedDescription
        }
    }

    @MainActor
    private func saveChanges() async {
        isSaving = true
        errorMessage = nil

        defer {
            isSaving = false
        }

        do {
            let allergensToAdd =
                selectedAllergenIds.subtracting(originalAllergenIds)

            let allergensToRemove =
                originalAllergenIds.subtracting(selectedAllergenIds)

            let restrictionsToAdd =
                selectedRestrictionIds.subtracting(originalRestrictionIds)

            let restrictionsToRemove =
                originalRestrictionIds.subtracting(selectedRestrictionIds)

            let preferencesToAdd =
                selectedPreferenceIds.subtracting(originalPreferenceIds)

            let preferencesToRemove =
                originalPreferenceIds.subtracting(selectedPreferenceIds)

            for id in allergensToAdd {
                try await DietaryService.addAllergen(id)
            }

            for id in allergensToRemove {
                try await DietaryService.removeAllergen(id)
            }

            for id in restrictionsToAdd {
                try await DietaryService.addDietaryRestriction(id)
            }

            for id in restrictionsToRemove {
                try await DietaryService.removeDietaryRestriction(id)
            }

            for id in preferencesToAdd {
                try await DietaryService.addDietaryPreference(id)
            }

            for id in preferencesToRemove {
                try await DietaryService.removeDietaryPreference(id)
            }

            dismiss()

        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

File: Android/app/src/main/java/com/pushtomaindev/kinkitchen/features/profile/EditDietaryProfileView.kt
package com.pushtomaindev.kinkitchen.features.profile

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.Role
import com.pushtomaindev.kinkitchen.components.buttons.KinPrimaryButton
import com.pushtomaindev.kinkitchen.components.feedback.KinLoadingView
import com.pushtomaindev.kinkitchen.components.labels.KinChip
import com.pushtomaindev.kinkitchen.components.labels.KinFlowLayout
import com.pushtomaindev.kinkitchen.components.navigation.KinNavigationBar
import com.pushtomaindev.kinkitchen.services.supabase.DietaryService
import com.pushtomaindev.kinkitchen.ui.theme.*
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch

/** One selectable lookup row: id plus display name. */
private data class DietaryItem(val id: String, val name: String)

/** Mirrors iOS `EditDietaryProfileView`. */
@Composable
fun EditDietaryProfileView(
    modifier: Modifier = Modifier,
    onDismiss: () -> Unit,
) {
    val scope = rememberCoroutineScope()

    var allergens by remember { mutableStateOf<List<DietaryItem>>(emptyList()) }
    var restrictions by remember { mutableStateOf<List<DietaryItem>>(emptyList()) }
    var preferences by remember { mutableStateOf<List<DietaryItem>>(emptyList()) }

    var selectedAllergens by remember { mutableStateOf<Set<String>>(emptySet()) }
    var selectedRestrictions by remember { mutableStateOf<Set<String>>(emptySet()) }
    var selectedPreferences by remember { mutableStateOf<Set<String>>(emptySet()) }

    // iOS keeps the loaded selection so saving can diff against it.
    var originalAllergens by remember { mutableStateOf<Set<String>>(emptySet()) }
    var originalRestrictions by remember { mutableStateOf<Set<String>>(emptySet()) }
    var originalPreferences by remember { mutableStateOf<Set<String>>(emptySet()) }

    var isLoading by remember { mutableStateOf(true) }
    var isSaving by remember { mutableStateOf(false) }
    var errorMessage by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(Unit) {
        isLoading = true
        errorMessage = null
        try {
            coroutineScope {
                val a = async { DietaryService.fetchAllergens() }
                val r = async { DietaryService.fetchDietaryRestrictions() }
                val p = async { DietaryService.fetchDietaryPreferences() }
                val sa = async { DietaryService.fetchSelectedAllergens() }
                val sr = async { DietaryService.fetchSelectedDietaryRestrictions() }
                val sp = async { DietaryService.fetchSelectedDietaryPreferences() }

                allergens = a.await().map { DietaryItem(it.id, it.name) }
                restrictions = r.await().map { DietaryItem(it.id, it.name) }
                preferences = p.await().map { DietaryItem(it.id, it.name) }

                selectedAllergens = sa.await().toSet()
                selectedRestrictions = sr.await().toSet()
                selectedPreferences = sp.await().toSet()

                originalAllergens = selectedAllergens
                originalRestrictions = selectedRestrictions
                originalPreferences = selectedPreferences
            }
        } catch (e: Exception) {
            errorMessage = e.message ?: "Unable to load your dietary options."
        } finally {
            isLoading = false
        }
    }

    fun saveChanges() {
        if (isSaving) return
        scope.launch {
            isSaving = true
            errorMessage = null
            try {
                // iOS writes only the difference, so untouched rows are left alone.
                (selectedAllergens - originalAllergens).forEach { DietaryService.addAllergen(it) }
                (originalAllergens - selectedAllergens).forEach { DietaryService.removeAllergen(it) }

                (selectedRestrictions - originalRestrictions).forEach {
                    DietaryService.addDietaryRestriction(it)
                }
                (originalRestrictions - selectedRestrictions).forEach {
                    DietaryService.removeDietaryRestriction(it)
                }

                (selectedPreferences - originalPreferences).forEach {
                    DietaryService.addDietaryPreference(it)
                }
                (originalPreferences - selectedPreferences).forEach {
                    DietaryService.removeDietaryPreference(it)
                }

                onDismiss()
            } catch (e: Exception) {
                errorMessage = e.message ?: "Unable to save your dietary profile."
            } finally {
                isSaving = false
            }
        }
    }

    Column(modifier.fillMaxSize().background(KinColors.background)) {
        KinNavigationBar("Edit Dietary Profile", showBackButton = true, onBack = onDismiss)

        if (isLoading) {
            Box(Modifier.fillMaxSize()) { KinLoadingView() }
        } else {
            Column(
                verticalArrangement = Arrangement.spacedBy(KinSpacing.xLarge),
                modifier = Modifier
                    .fillMaxSize()
                    .verticalScroll(rememberScrollState())
                    .padding(KinSpacing.xLarge),
            ) {
                SelectableSection("Allergens", allergens, selectedAllergens) { id ->
                    selectedAllergens = selectedAllergens.toggled(id)
                }
                SelectableSection("Dietary Restrictions", restrictions, selectedRestrictions) { id ->
                    selectedRestrictions = selectedRestrictions.toggled(id)
                }
                SelectableSection("Dietary Preferences", preferences, selectedPreferences) { id ->
                    selectedPreferences = selectedPreferences.toggled(id)
                }

                errorMessage?.let {
                    Text(it, style = KinTypography.body, color = KinColors.error)
                }

                KinPrimaryButton(
                    title = if (isSaving) "Saving..." else "Save Changes",
                    color = KinColors.success,
                    enabled = !isSaving,
                ) { saveChanges() }
            }
        }
    }
}

@Composable
private fun SelectableSection(
    title: String,
    items: List<DietaryItem>,
    selectedIds: Set<String>,
    onToggle: (String) -> Unit,
) {
    Column(
        verticalArrangement = Arrangement.spacedBy(KinSpacing.medium),
        modifier = Modifier.fillMaxWidth(),
    ) {
        Text(title, style = KinTypography.title, color = KinColors.primaryText)

        KinFlowLayout(spacing = KinSpacing.small) {
            items.forEach { item ->
                KinChip(
                    title = item.name,
                    isSelected = item.id in selectedIds,
                    modifier = Modifier.clickable(role = Role.Checkbox) { onToggle(item.id) },
                )
            }
        }
    }
}

private fun Set<String>.toggled(id: String): Set<String> =
    if (id in this) this - id else this + id
